subset/security/password/run_password_test_for_protocol

```diff
@@ -0,0 +1,165 @@
+#!/bin/bash
+
+# A test which runs a dictionary brute force attack using ncrack/medusa, saving the result into a file afterwards.
+#
+# Supports http, https, telnet and ssh protocols on their specified ports.
+#
+# Assumes the following files are in the dictionary directory:
+# - dictionary.txt (File with format ":username:password")
+# - usernames.txt (File with a list of usernames, one on each line)
+# - passwords.txt (File with a list of passwords, one on each line)
+#
+# Usage: ./run_password_test_for_protocol $TARGET_IP, $PROTOCOL, $PORT, $DICTIONARY_DIR, $RESULTS_DIR
+
+TARGET_IP=$1
+PROTOCOL=$2
+PORT=$3
+DICTIONARY_DIR=$4
+RESULT_DIR=$5
+
+DICTIONARY="$DICTIONARY_DIR/dictionary.txt"
+USERNAMES="$DICTIONARY_DIR/usernames.txt"
+PASSWORDS="$DICTIONARY_DIR/passwords.txt"
+RESULT_FILE="${RESULT_DIR}/security_password_${PROTOCOL}.result"
+
+# Runs the nmap on a target host and port. Returns the output of the command.
+# $1 Target IP
+# $2 Target Port
+function run_nmap_and_get_command() {
+  NMAP_OUTPUT=$(nmap -p $2 $1)
+  echo "$NMAP_OUTPUT"
+}
+
+# Return true if grep finds a match for the nmap port open message which looks like: e.g. "80/tcp open http"
+# $1 Nmap command output
+# $2 Target protocol
+# $3 Target port
+function is_specified_port_open() {
+  echo "$1" | grep -q "$3.*open.*$2"
+}
+
+# Returns true if grep can find the host down message from the Nmap output
+# $1 Nmap command output
+function is_host_down() {
+  echo "$1" | grep -q "Host seems down"
+}
+
+# Checks if http or https has authentication.
+# $1 Target protocol (HTTP or HTTPS)
+# $2 Target IP
+# $3 Target port
+function does_http_or_https_have_authentication() {
+  if [[ $1 == "http" || $1 == "https" ]]; then
+    echo "$(curl -k --silent $1://$2:$3)" | grep -q "401 Authorization Required"
```

Comment:
@em-redstone  you should look in the headers for 401 response code rather than the HTML body, as the text varies between servers. For example Apache will instead say “401 Unauthorized”, which results in a test skip. You could add the '-i' flag to the curl request

Reply:
Hi @noursaidi,

I've added the changes - It now only looks in the headers for the response code.

I've also heard mention in the meeting, upcoming features (operator to decide which password file to use) to be added to the password module - Will that be the only feature to be added, or should more be expected to come? Also who will now be expected to implement them?

Many thanks.

Reply:
Thanks @em-redstone, looks good and it all works now as expected. Thanks for incorporating this comment and all the other comments we've had.

I think the ability for a test operator to provide a username & password combination is the final feature which is needed and completes the test, I can't see any other changes required currently. I'm happy to incorporate this additional functionality

Reply:
With regards to this feature, the module recently now has the option to specify which files the password test will use in the module_config.json (Please see readme.md for more info) so I actually think it's been implemented already.

Currently there are two sub folders within the resource folder, which is copied into the docker container.
- faux: Which contains a shortened version of the password list for faux device testing. The script testing/test_aux.sh updates the config to use this folder instead of the default.
- default: The full password list currently used by the module.

If the operator needs to add their own - ideally they will just need to make a new folder in there with their own version the files, and update the module_config.json

The files needed would be:
- A .txt file containing the usernames and password pairs in the format :username:password
- Two separate files containing the same usernames and passwords in the above list.

Reason for the multiple files is that ncrack doesn't seem to support colon separated files.

Many thanks.

Reply:
Thanks @em_redstone, I noticed that change when looking at the code, and is a very useful and welcome if someone wanted a more tailored password list and to deal with the length of the test. The change we had in mind was to expand on it, instead of specifying a file they specify the username and password directly along with the device serial, etc, in module_config

```
“default_username”: “*** (optional) ***”,
“default_password”: “*** (optional) ***”
```

Then, in test_password you could have these into a new dictionary with a single line. I haven't tested this, but it i hope it explains what I had in mind.

```
USERNAME_KEY=".device_info.default_username"
PASSWORD_KEY=".device_info.default_password"

DEFAULT_USERNAME="$(get_module_config_value_from_key $MODULE_CONFIG $USERNAME_KEY)"
DEFAULT_PASSWORD="$(get_module_config_value_from_key $MODULE_CONFIG $PASSWORD_KEY)"

if [[ -z "${$DEFAULT_USERNAME}" ]]
    echo "${DEFAULT_USERNAME}:${DEFAULT_PASSWORD}:" >> provided_credentials.txt
    NEW_DICTIONARY_DIR = provided_credentials.txt
fi 
```

Reply:
Hi @noursaidi,

Will the other feature mentioned in the spec sheet involving the output of the different protocols to be merged into a single result be required?

Many thanks.

Reply:
@em-redstone I believe when the tests were renamed recently they were retained as separate tests so I thought not. @pisuke @spectre-storm please confirm?

Reply:
Hi @noursaidi,

Please take a look at the recent code change.

The module can now use the specified "default_username" and "default_password" value in the module config to use as the only combination when brute forcing.

The process added involves creating another folder called "resource/user_specified" and its constituent files "dictionary.txt", "usernames.txt" and "passwords.txt" with the username/password specified as the only values.

Readme.md for this file has also been updated.

Many thanks.